doc/treeconfig.md
# Tree Configuration File

This is the default tree configuration file.

    Automatic Destruction:

    Automatic Destruction:

    # Main switch to deactivate automatic destruction
    Active: true

    # Damage the player's tool for every block of the tree, not just the first they broke
    Apply Full Tool Damage: true

    # Damage the player's tool for leaves broken automatically
    Apply Tool Damage For Leaves: true

    # Add the logs the player's inventory
    Auto Add To Inventory: false

    # Seconds to wait before (force) removing remnants of the tree
    Cleanup Delay Seconds: 20

    # If we clean up remnants, we also will remove leaves without drops
    Cleanup Leaves: true

    # Time to wait before allowing the player to automatically destroy again
    Cooldown Seconds: 0

    # Ticks to wait before breaking the next block, set to -1 for instant breaking
    Delay Ticks: 0

    # Always remove remnants of the tree, as soon as a tree has been verified and is being broken
    Forced Removal: false

    # Main switch for the Block Statistic nodes
    Increases Statistics: false

    # Initial Delay before actually starting to break the tree
    Initial Delay: false

    # Seconds to delay automatic destruction
    Initial Delay Seconds: 10

    # Remove not only logs, but also leaves
    Remove Leaves: true

    # Required lore on tool in order to automatically remove a tree. Empty means no requirement
    Required Lore: ''

    # Only automatically destroy with the right tools, they are set in the tree definitions or via command
    Requires Tools: true

    # Automatically destroy when sneaking
    When Sneaking: true

    # Automatically destroy when not sneaking
    When Not Sneaking: true

    # Support silk touch affect when a player has it
    Use Silk Touch: true
    Block Statistics:

    # Count minecraft block breaking statistics when automatically breaking
    Mine Block: false

    # Count minecraft pickup statistics when automaticall adding blocks to inventory
    Pickup: false
    Blocks:
    Cap:

        # Max height of a branch cap
        Height: 2

        # Max radius of a branch cap
        Radius: 3
    Custom Drops:

        # Generate custom drops according to the list
        Active: true

        # Drop chances for extra drops. 1.0 would be 100% chance!
        Items:
          minecraft:apple: 0.01
          minecraft:golden_apple: 0.001

        # These are additional factors, for example, by default, iron has half the chance to get custom drops
        Factors:
          minecraft:netherite_axe: 1.0
          minecraft:diamond_axe: 1.0
          minecraft:golden_axe: 0.75
          minecraft:iron_axe: 0.5
          minecraft:stone_axe: 0.25
          minecraft:wooden_axe: 0.1

        # Custom Drops below completely replace the leaf drops
        Override: false

    # Here you can add extra blocks that can be expected inside or around tree leaves
    Materials:
    - 'minecraft:bee_nest'
      Middle:

      # Allow air pockets in leaves
      Air: false

      # Check cubic edges
      Edges: false

      # Radius around the trunk to check for leaves
      Radius: 2

    # How many leaves do we require for it to be a valid tree
    Required: 10
    Top:

        # Allow air pockets in leaves
        Air: false

        # Check cubic edges
        Edges: false

        # Radius around the trunk to check for leaves
        Radius: 3

        # Height above the trunk to check for leaves
        Height: 3

    # Valid blocks that are below and around the saplings
    Ground Blocks:
    - 'minecraft:podzol'
    - 'minecraft:mycelium'
    - 'minecraft:grass_block'
    - 'minecraft:dirt'
    - 'minecraft:coarse_dirt'
    - 'minecraft:sand'

    # Blocks that you can expect to be around the tree - these are the exceptions from player building safeguards
    Natural Blocks:
    - 'minecraft:air'
    - 'minecraft:water'
    - 'minecraft:fire'
    - 'minecraft:snow'
    - 'minecraft:snow_block'
    - 'minecraft:stone'
    - 'minecraft:grass_block'
    - 'minecraft:dirt'
    - 'minecraft:coarse_dirt'
    - 'minecraft:sand'
    - 'minecraft:terracotta'
    - '*_terracotta'
    - 'minecraft:mycelium'
    - 'minecraft:podzol'
    - '*_sapling'
    - '*_leaves'
    - 'minecraft:dandelion'
    - '*_tulip'
    - 'minecraft:poppy'
    - '*_orchid'
    - 'minecraft:allium'
    - 'minecraft:azure_bluet'
    - 'minecraft:oxeye_daisy'
    - 'minecraft:lilac'
    - 'minecraft:short_grass'
    - 'minecraft:tall_grass'
    - 'minecraft:fern'
    - 'minecraft:dead_bush'
    - 'minecraft:sugar_cane'
    - '*_mushroom_block'
    - 'minecraft:mushroom_stem'
    - 'minecraft:melon'
    - 'minecraft:pumpkin'
    - 'minecraft:bee_nest'
    - 'minecraft:cocoa'
    - 'minecraft:torch'
    - 'minecraft:rail'
    - 'minecraft:hopper'
    - 'minecraft:oak_sapling'
    - 'minecraft:birch_sapling'
    - 'minecraft:spruce_sapling'
    - 'minecraft:jungle_sapling'
    - 'minecraft:dark_oak_sapling'
    - 'minecraft:acacia_sapling'
    - 'minecraft:bamboo'
    - 'minecraft:vine'
    - 'minecraft:cave_air'
    - 'minecraft:void_air'
    - 'minecraft:large_fern'
    - 'minecraft:peony'
    - 'red_mushroom'
    - 'brown_mushroom'

    # The permission required for this tree type
    Permission: treeassist.destroy.*
    Replanting:

    # Main switch to deactivate sapling replanting
    Active: true

    # How far down do we look for saplings to place them
    Check Depth: 12

    # How long to wait before placing a sapling. Should stay above 0 because of bukkit event handling
    Delay: 1
    Dropped:

        # Attempt to plant a dropped sapling item
        Active: false

        # What is the chance for us doing this, 1.0 means 100%
        Probability: 0.1

        # How many ticks should we wait until attempting to plant it
        Delay Ticks: 5

    # Even if something would prevent sapling replacement or auto destruction, we will place a sapling
    Enforce: false

    # Prevent from breaking this type of sapling at all costs
    Force Protect: false

    # How long should saplings stay there before they can grow
    Growth Delay Seconds: 0

    # The material to place
    Material: minecraft:air

    # Only place saplings when the bottom block was broken
    Only When Bottom Block Broken First: false

    # How long to protect saplings
    Protect For Seconds: 0

    # Only replant with the right tools, they are set in the tree definitions or via command
    Requires Tools: true

    # Replant when a tree block burns
    When Tree Burns Down: true

    # This is the list that can be required to use when auto destructing or sapling replanting
    Tool List:
    - 'minecraft:netherite_axe'
    - 'minecraft:diamond_axe'
    - 'minecraft:wooden_axe'
    - 'minecraft:golden_axe'
    - 'minecraft:iron_axe'
    - 'minecraft:stone_axe'
      Trunk:

      # Look for branches
      Branch: false
      Custom Drops:

      # Generate custom drops
      Active: false

      # Drop chances for extra drops. 1.0 would be 100% chance!
      Items:
      minecraft:golden_apple: 0.0001

      # These are additional factors, for example, by default, iron has half the chance to get custom drops
      Factors:
      minecraft:netherite_axe: 1.0
      minecraft:diamond_axe: 1.0
      minecraft:golden_axe: 0.75
      minecraft:iron_axe: 0.5
      minecraft:stone_axe: 0.25
      minecraft:wooden_axe: 0.1

      # The configured drops override regular drops, including logs!
      Override: false

      # The trunk can go diagonally
      Diagonal: false

      # How high does it need to be to qualify as a tree
      Minimum Height: 4

      # How thick is the trunk
      Thickness: 1

      # Saplings can be at different height
      Uneven Bottom: false

    # Version number for automagical config updates
    Version: 7.3019

doc/configuration.md
# Main Configuration File

This is the default main configuration file.

    General:
      # Language file to load. Does not need YML extension!
      Language: lang_en
      # Prevent breaking of logs without the right tool
      Prevent Log Breaking Without Tool: false
      # For the toggling command, should players start with TreeAssist active?
      Toggle Default: true
      # Should the plugin remember the toggle states on restart?
      Toggle Remember: false
      # Should we check if players have permissions? If false, all features are available to everyone.
      Use Permissions: false
      # Should we announce information about every tree config we load?
      Verbose Config Loading: true
    # === [ Command Settings ] ===
    Commands:
      Force Break:
        Default Radius: 10
        Max Radius: 30
      Force Grow:
        Default Radius: 10
        Max Radius: 30
      No Replant:
        Cooldown Time: 30
    # === [ Automatic Destruction Settings ] ===
    Destruction:
      # Spawn a FallingBlock when breaking a block
      Falling Blocks: false
      # Make the FallingBlocks look almost like an actual tree falling over
      Falling Blocks Fancy: false
      # Increase leaf decay by looking for nearby lonely leaves
      Fast Leaf Decay: true
      # Only break blocks that are above the block the player broke
      Only Above: false
    # === [ Placed Blocks Settings ] ===
    Placed Blocks:
      # Check for player placed blocks
      Active: true
      # A plugin that will look for placed blocks (Prism, LogBlock, CoreProtect)
      Plugin Name: TreeAssist
      # How many seconds back we want to look
      Lookup Time: 86400
    # === [ Plugin Integration Settings ] ===
    Plugins:
      # Count broken blocks towards the TreeFeller ability
      mcMMO: true
      # Count broken trees towards the TreeFeller ability
      TreeMcMMO: false
      # Count broken blocks towards Jobs jobs that fit
      Jobs: true
      # Count broken trees towards Jobs jobs that fit
      TreeJobs: false
      # Allow to set up regions with WorldGuard to prevent destruction with 'treeassist-autochop' and sapling replacement with 'treeassist-replant'
      WorldGuard: false
    # === [ World Related Settings ] ===
    Worlds:
      # Enable world based restrictions
      Restrict: false
      # Worlds that are not restricted
      Enabled Worlds:
      - world
      - world2
    Modding:
      # This is for hacky mods that use infinity durability
      Disable Durability Fix: false
    Debug: none
    # === [ bStats Metrics Settings ] ===
    bStats:
      # Should we send stats at all? Please keep this in so we have an overview of spread of versions <3
      Active: true
      # This setting sends a bit more detailed information about which features are used at all. Thank you for supporting me!
      Full: true
    # === [ Updater Settings ] ===
    Update:
      # what to do? Valid values: disable, announce, download, both
      Mode: both
      # which type of branch to get updates? Valid values: dev, alpha, beta, release
      Type: beta
    # The config version for update checks
    Version: 7.3027

doc/changelog.md
# v7.X Changelog

## v7.3 - API Expansion

- v7.3.51 - update for 1.20.4, also make updating minecraft-version-dependent to not break servers by introducing unknown blocks or tree definitions
- v7.3.50 - print correct permission message for "treeconfig" command
- v7.3.49 - implement pull #97 manually
- v7.3.48 - allow natural blocks to be the roof of a tree - nether hotfix
- v7.3.47 - add language entries for on and off
- v7.3.46 - add placeholders ta_global - ta_world - ta_cooldown - for toggle and cooldown status
- v7.3.45 - /ta config restore [config] | Restore a tree config to its default
- v7.3.44 - help with cherry tree support by allowing pink petals near the tree
- v7.3.43 - update locate biome message to reflect the new command since 1.19
- v7.3.42 - allow restricting max dimensions of a tree, this counts the trunk only
- v7.3.41 - support cherry trees
- v7.3.40 - change how the toggle check command works, allow to check for other players now though
- v7.3.39 - allow to check your toggle status by appending "check" after the world, or the toggle command
- v7.3.38 - implement adding all tree block drops to inventory (can be disabled)
- v7.3.37 - fix an Aurelium related config warning, addresses github issue #90
- v7.3.36 - help with AureliumSkills compatibility by waiting for it to load before hooking into it
- v7.3.35 - stop the auto plant feature from duplicating saplings
- v7.3.34 - repair several things we broke in the last few commits
- v7.3.33 - actually add the config setting, automatically. Oups
- v7.3.32 - add config setting to ignore invalid blocks - use with caution :P - add aurelium skills auto-addition
- v7.3.31 - repair and improve the tree yml replacement
- v7.3.30 - add the mangrove to the automatically added config files, repair the add-config-feature
- v7.3.29 - follow greedy branches downwards, too
- v7.3.28 - try to implement mangroves - I am not completely happy with this. Config will not be automatically added yet!
- v7.3.27 - add a config setting to allow to remember toggle settings -> General.ToggleRemember
- v7.3.26 - add a whole new world of trunk discovery. greedy. use with caution!
- v7.3.25 - add french language file. Thanks to TBG-FR!
- v7.3.24 - actually respect language setting! This has been broken for a long time :(
- v7.3.23 - add AureliumSkills support
- v7.3.22 - purge language values no longer in use, add english and german language file
- v7.3.21 - fix a few issues resulting in enforcing of removal, and fix someone probably being drunk while implementing the statistic check
- v7.3.20 - improve debugging when the plugin is toggled off globally. It happens ^^
- v7.3.19 - improve cleanup by allowing to not clean up leaves, set "AutomaticDestruction.CleanupLeaves" to "false" to allow leaves to drop items even with cleanup
- v7.3.18 - improve "info" parameter for config commands, it now works properly and is readable, treeconfig no longer requires tree type for node info
- v7.3.17 - move Replanting config entry "Delay Growth Seconds" to "Growth Delay Seconds" to allow for config command arguments to be unambiguous
- v7.3.16 - plot twist! We can actually replace a sapling without auto destruct IF the first broken block is the bottom block!
- v7.3.15 - add nyliums to the valid mushroom ground blocks
- v7.3.14 - move to HTTPS for Updater [core update], show external plugin stats more realistically
- v7.3.13 - add URL verification to prevent tampering with it [core update]
- v7.3.12 - do not drop blocks from falling blocks when we are cleaning up
- v7.3.11 - only lock player's tool when we want to do damage later
- v7.3.10 - fix NPE in the "prevent without tool"  logic
- v7.3.9 - properly handle trees without leaves
- v7.3.8 - skip null trunk tree removal (why is this even happening?)
- v7.3.7 - call the TreeBrokenEvent with higher priority, extend TreeStructure API
- v7.3.6 - synchronize version numbers
- v7.3.1 - expand TreeBrokenEvent, add TreeBlockBrokenEvent and call both in the appropriate places

## v7.2 - Library

- v7.2.53 - address ticket 0062 - skip item spawn events of spammy items that are no saplings
- v7.2.52 - address ticket 0063 - tweak check for self-replant, add config for different tree types
- v7.2.51 - address ticket 0061 - noreplant did not take proper priority. Also, now noreplant and replant will trump each other on command
- v7.2.50 - add debug command parameter to get spammed about why things are not working, more in the docs!
- v7.2.49 - prevent players dropping their axes to avoid tool damage
- v7.2.48 - allow acacia and potentially other trees to actually remove cap leaves, improve thick tree determination for even less infringement
- v7.2.47 - add a debugtool command to show the extent of trees, fix thick jungle trees stealing leaves from other jungle trees
- v7.2.46 - add github issue #77 - "Prevent Log Breaking Without Tool" - it will check before trashing your tool when auto destructing
- v7.2.45 - add github issue #71 - /treeassist replant - to enforce sapling planting for yourself for a limited time
- v7.2.44 - support tree farms below 0 [y position]
- v7.2.43 - fast decaying leaves will drop items now - can be disabled in the main config
- v7.2.42 - use proper Override annotation for onCommand
- v7.2.41 - Bye bye, biomes [1.18]
- v7.2.40 - prevent houses built out of logs to count as farming rows
- v7.2.39 - do the legacy loading in the correct order to populate nonexisting tree ymls
- v7.2.38 - try to repair the legacy 7.0 config fix
- v7.2.37 - fix an issue with tree growing triggering the block protection
- v7.2.36 - prevent NPE when running the command with no arguments
- v7.2.35 - support hex color codes (Core Update)
- v7.2.34 - properly implement "Only Above" logic. That "above" should be inclusive so that it does not leave the base level unbroken. Addresses #80
- v7.2.33 - include azalea in the .* permission
- v7.2.32 - add commands per block / per tree, documentation will follow soon!

src/main/java/me/itsatacoshop247/TreeAssist/core/Config.java
package me.itsatacoshop247.TreeAssist.core;

import org.bukkit.Material;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.util.*;

public class Config {
    private final YamlConfiguration cfg;
    private final File configFile;
    private final Map<String, Boolean> booleans;
    private final Map<String, Integer> ints;
    private final Map<String, Double> doubles;
    private final Map<String, String> strings;

    public void save() {
        try {
            cfg.save(configFile);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public enum CFG {
        MAIN_DESTROY_ONLY_BLOCKS_ABOVE("Main.Destroy Only Blocks Above", false),
        MAIN_FORCE_BREAK_DEFAULT_RADIUS("Main.Force Break Default Radius", 10),
        MAIN_FORCE_GROW_DEFAULT_RADIUS("Main.Force Grow Default Radius", 10),
        MAIN_FORCE_BREAK_MAX_RADIUS("Main.Force Break Max Radius", 30),
        MAIN_FORCE_GROW_MAX_RADIUS("Main.Force Grow Max Radius", 30),
        MAIN_IGNORE_USER_PLACED_BLOCKS("Main.Ignore User Placed Blocks", false),
        MAIN_LANGUAGE("Main.Language", "en"),
        MAIN_TOGGLE_DEFAULT("Main.Toggle Default", true),
        MAIN_USE_MCMMO_IF_AVAILABLE("Main.Use mcMMO if Available", true),
        MAIN_USE_JOBS_IF_AVAILABLE("Main.Use Jobs if Available", true),
        MAIN_USE_WORLDGUARD_IF_AVAILABLE("Main.Use WorldGuard if Available", false),
        MAIN_USE_PERMISSIONS("Main.Use Permissions", false),
        MAIN_USE_FALLING_BLOCKS("Main.Use Falling Blocks", false),

        AUTOMATIC_TREE_DESTRUCTION_WHEN_SNEAKING("Automatic Tree Destruction.When Sneaking", true),
        AUTOMATIC_TREE_DESTRUCTION_WHEN_NOT_SNEAKING("Automatic Tree Destruction.When Not Sneaking", true),

        LEAF_DECAY_FAST_LEAF_DECAY("Leaf Decay.Fast Leaf Decay", true),

        SAPLING_REPLANT_BLOCK_ALL_BREAKING_OF_SAPLINGS("Sapling Replant.Block all breaking of Saplings", false),
        SAPLING_REPLANT_BOTTOM_BLOCK_HAS_TO_BE_BROKEN_FIRST("Sapling Replant.Bottom Block has to be Broken First", true),
        SAPLING_REPLANT_COMMAND_TIME_DELAY("Sapling Replant.Command Time Delay (Seconds)", 30),

        SAPLING_REPLANT_REPLANT_WHEN_TREE_BURNS_DOWN("Sapling Replant.Replant When Tree Burns Down", true),
        SAPLING_REPLANT_TIME_TO_PROTECT_SAPLING("Sapling Replant.Time to Protect Sapling (Seconds)", 0),
        SAPLING_REPLANT_TIME_TO_BLOCK_SAPLING_GROWTH("Sapling Replant.Time to Block Sapling Growth (Seconds)", 0),

        WORLDS_ENABLE_PER_WORLD("Worlds.Enable Per World", false),
        WORLD_ENABLED_WORLDS("Worlds.Enabled Worlds", new ArrayList<>(Arrays.asList("world", "world2"))),

        PLACED_BLOCKS_HANDLER_PLUGIN_NAME("Placed Blocks.Handler Plugin Name", "TreeAssist"),
        PLACED_BLOCKS_HANDLER_LOOKUP_TIME("Placed Blocks.Handler Lookup Time", 86400),

        MODDING_DISABLE_DURABILITY_FIX("Modding.Disable Durability Fix", false),

        VERSION("Version", 6.0);



        private final String node;
        private final Object value;
        private final String type;

        public static CFG getByNode(final String node) {
            for (final CFG m : CFG.getValues()) {
                if (m.node.equals(node)) {
                    return m;
                }
            }
            return null;
        }

        CFG(final String node, final String value) {
            this.node = node;
            this.value = value;
            type = "string";
        }

        CFG(final String node, final Boolean value) {
            this.node = node;
            this.value = value;
            type = "boolean";
        }

        CFG(final String node, final Integer value) {
            this.node = node;
            this.value = value;
            type = "int";
        }

        CFG(final String node, final Double value) {
            this.node = node;
            this.value = value;
            type = "double";
        }

        CFG(final String node, final List<String> value) {
            this.node = node;
            this.value = value;
            this.type = "list";
        }

        public String getNode() {
            return node;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        public Object getValue() {
            return value;
        }

        public static CFG[] getValues() {
            return values();
        }

        public String getType() {
            return type;
        }

    }


    /**
     * Create a new TreeConfig instance that uses the specified file for loading
     *
     * @param configFile a YAML file
     */
    public Config(final File configFile) {
        Utils.plugin.getLogger().info("Loading tree config file: " + configFile.getAbsolutePath().replace(Utils.plugin.getDataFolder().getAbsolutePath(), ""));

        cfg = new YamlConfiguration();
        this.configFile = configFile;

        if (ConfigUpdater.check(cfg)) {
            save();
        }

        booleans = new HashMap<>();
        ints = new HashMap<>();
        doubles = new HashMap<>();
        strings = new HashMap<>();
    }

    /**
     * Load the config-file into the YamlConfiguration, and then populate the
     * value maps.
     *
     * @return true, if the load succeeded, false otherwise.
     */
    public boolean load() {
        try {
            cfg.load(configFile);
            reloadMaps();
            return true;
        } catch (final Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Iterates through all keys in the config-file, and populates the value
     * maps. Boolean values are stored in the booleans-map, Strings in the
     * strings-map, etc.
     */
    public void reloadMaps() {
        // known exceptions
        String[] exceptions = {"Main",
                "Automatic Tree Destruction", "Block Statistics", "Leaf Decay",
                "Sapling Replant", "Worlds", "Placed Blocks"};

        root: for (final String s : cfg.getKeys(true)) {
            final Object object = cfg.get(s);

            if (object instanceof Boolean) {
                booleans.put(s, (Boolean) object);
            } else if (object instanceof Integer) {
                ints.put(s, (Integer) object);
            } else if (object instanceof Double) {
                doubles.put(s, (Double) object);
            } else if (object instanceof String) {
                strings.put(s, (String) object);
            }

            if (CFG.getByNode(s) == null) {

                for (String test : exceptions) {
                    if (s.equals(test)) {
                        continue root;
                    }
                }

                Utils.plugin.getLogger().warning("No valid node: " + s);
            }
        }
    }


    // /////////////////////////////////////////////////////////////////////////
    // //
    // GETTERS //
    // //
    // /////////////////////////////////////////////////////////////////////////

    /**
     * Get the YamlConfiguration associated with this Config instance. Note that
     * changes made directly to the YamlConfiguration will cause an
     * inconsistency with the value maps unless reloadMaps() is called.
     *
     * @return the YamlConfiguration of this Config instance
     */
    public YamlConfiguration getYamlConfiguration() {
        return cfg;
    }

    /**
     * Retrieve a value from the YamlConfiguration.
     *
     * @param string the path of the value
     * @return the value of the path
     */
    public Object getUnsafe(final String string) {
        return cfg.get(string);
    }

    /**
     * Retrieve a boolean from the value maps.
     *
     * @param cfg the node of the value
     * @return the boolean value of the path if the path exists, false otherwise
     */
    public boolean getBoolean(final CFG cfg) {
        return getBoolean(cfg, (Boolean) cfg.getValue());
    }

    /**
     * Retrieve a boolean from the value maps.
     *
     * @param cfg the node of the value
     * @param def a default value to return if the value was not in the map
     * @return the boolean value of the path if it exists, def otherwise
     */
    private boolean getBoolean(final CFG cfg, final boolean def) {
        final String path = cfg.getNode();
        final Boolean result = booleans.get(path);
        return result == null ? def : result;
    }

    /**
     * Retrieve an int from the value maps.
     *
     * @param cfg the node of the value
     * @return the int value of the path if the path exists, 0 otherwise
     */
    public int getInt(final CFG cfg) {
        return getInt(cfg, (Integer) cfg.getValue());
    }

    /**
     * Retrieve an int from the value maps.
     *
     * @param cfg the node of the value
     * @param def a default value to return if the value was not in the map
     * @return the int value of the path if it exists, def otherwise
     */
    public int getInt(final CFG cfg, final int def) {
        final String path = cfg.getNode();
        final Integer result = ints.get(path);
        return result == null ? def : result;
    }

    /**
     * Retrieve a double from the value maps.
     *
     * @param cfg the node of the value
     * @return the double value of the path if the path exists, 0D otherwise
     */
    public double getDouble(final CFG cfg) {
        return getDouble(cfg, (Double) cfg.getValue());
    }

    /**
     * Retrieve a double from the value maps.
     *
     * @param cfg the node of the value
     * @param def a default value to return if the value was not in the map
     * @return the double value of the path if it exists, def otherwise
     */
    public double getDouble(final CFG cfg, final double def) {
        final String path = cfg.getNode();
        final Double result = doubles.get(path);
        return result == null ? def : result;
    }

    /**
     * Retrieve a string from the value maps.
     *
     * @param cfg the node of the value
     * @return the string value of the path if the path exists, null otherwise
     */
    public String getString(final CFG cfg) {
        return getString(cfg, (String) cfg.getValue());
    }

    /**
     * Retrieve a string from the value maps.
     *
     * @param cfg the node of the value
     * @param def a default value to return if the value was not in the map
     * @return the string value of the path if it exists, def otherwise
     */
    public String getString(final CFG cfg, final String def) {
        final String path = cfg.getNode();
        final String result = strings.get(path);
        return result == null ? def : result;
    }

    /**
     * Retrieve a list of materials from the value maps
     *
     * @param cfg the node of the value
     * @return a list of materials (can contain null)
     */
    public List<Material> getMaterials(CFG cfg) {
        List<String> list = getStringList(cfg, new ArrayList<>());

        List<Material> matList = new ArrayList<>();

        for (String matName : list) {
            matList.add(Material.matchMaterial(matName));
        }

        return matList;
    }

    public Set<String> getKeys(final String path) {
        if (cfg.get(path) == null) {
            return null;
        }

        final ConfigurationSection section = cfg.getConfigurationSection(path);
        return section == null? new HashSet<>() : section.getKeys(false);
    }

    public List<String> getStringList(final CFG cfg, final List<String> def) {
        if (this.cfg.get(cfg.node) == null) {
            return def == null ? new LinkedList<>() : def;
        }

        return this.cfg.getStringList(cfg.node);
    }
}
